src/main/kotlin/com/hj/leetcode/kotlin/problem983/Solution.kt

```diff
@@ -0,0 +1,46 @@
+package com.hj.leetcode.kotlin.problem983
+
+/**
+ * LeetCode page: [983. Minimum Cost For Tickets](https://leetcode.com/problems/minimum-cost-for-tickets/);
+ */
+class Solution {
+    /* Complexity:
+     * Time O(N) and Space O(N) where N is the size of days;
+     */
+    fun mincostTickets(days: IntArray, costs: IntArray): Int {
+        // suffixMinCost[i] ::= the min cost of the suffix array of days start from index i
+        val suffixMinCost = IntArray(days.size + 1)
+        for (i in days.indices.reversed()) {
+            // Case 1: We buy a day pass on day[i]
+            val firstDay = days[i]
+            val dayPassMinCost =
+                costs[0] + suffixMinCost[days.firstIndex(fromIndex = i) { day -> day > firstDay }]
+            // Case 2: We buy a week pass on day[i]
+            val weekPassExpiryDay = firstDay + 6
+            val weekPassMinCost =
+                costs[1] + suffixMinCost[days.firstIndex(fromIndex = i) { day -> day > weekPassExpiryDay }]
```

Comment:
we can reuse the found fisrt index for suffix i as starting point and do a backward search to find the first index for suffix i-1

Reply:
Refer to #14 for the improved solution